sys/net/link_layer/ieee802154/submac.c

```diff
@@ -0,0 +1,384 @@
+#include <stdio.h>
+#include <string.h>
+#include "net/ieee802154/submac.h"
+#include "net/ieee802154.h"
+#include "xtimer.h"
+#include "random.h"
+#include "luid.h"
+#include "kernel_defines.h"
+#include "errno.h"
+#include <assert.h>
+
+#define CSMA_SENDER_BACKOFF_PERIOD_UNIT (320U)
+#define ACK_TIMEOUT                     (864U)
+
+#define IEEE802154_CCA_THRESH_DEFAULT   (-70)
+
+#if IS_USED(MODULE_IEEE802154_SUBMAC)
+
+static void _handle_tx_no_ack(ieee802154_submac_t *submac);
+
+static void _tx_end(ieee802154_submac_t *submac, int status,
+                    ieee802154_tx_info_t *info)
+{
+    ieee802154_dev_t *dev = submac->dev;
+
+    ieee802154_radio_request_set_trx_state(dev, submac->state == IEEE802154_STATE_LISTEN ? IEEE802154_TRX_STATE_RX_ON : IEEE802154_TRX_STATE_TRX_OFF);
+
+    submac->tx = false;
+    while (ieee802154_radio_confirm_set_trx_state(dev) == -EAGAIN) {}
+    submac->cb->tx_done(submac, status, info);
+}
+
+static inline bool _does_handle_ack(ieee802154_dev_t *dev)
+{
+    return ieee802154_radio_has_frame_retrans(dev) ||
+           ieee802154_radio_has_irq_ack_timeout(dev);
```

Comment:
yes! In fact, this was described in the original RDM. I tried it in some branch and the code size can be reduced quite a lot.

Reply:
The idea was to define the `ieee802154_radio_has_xxx` functions like this:

```c
static inline bool ieee802154_radio_has_frame_retrans(ieee802154_t *dev)
{
    if (IS_ACTIVE(IEEE802154_SUBMAC_FRAME_RETRANS)) { 
        if (IS_ACTIVE(IEEE802154_RADIO_HAS_FRAME_RETRANS) {
            /* If the SubMAC handles retransmissions and there's at least one radio with the cap, use the `get_cap` function */
            return dev->driver->get_cap(dev, IEEE802154_CAP_FRAME_RETRANS);
        }
        /* Otherwise, assume all radios require software frame retransmission.
        return false;
    }
   else {
        /* If the SubMAC doesn't take care of the frame retransmissions, assume ell radios provide hardware frame retransmissions */
        assert(IS_ACTIVE(IEEE802154_RADIO_HAS_FRAME_RETRANS));
        return true;
    }
}
```

That way, we don't have to add guards in all SubMAC functions.

Reply:
I quickly tested it with `cc2538dk`.
From:
```
   text	  data	   bss	   dec	   hex
  22660	   632	  3748	 27040	  69a0	
```

Goes to:
```
 text	  data	   bss	   dec	   hex
  22248	   632	  3748	 26628	  6804
```

Reply:
are you ok with leaving these optimizations for follow up PRs? most of them will become even easier when Kconfig can resolve dependencies

Reply:
Yea sure, that's what I did for the sub-modes of `at86rf215` as well.
Just wanted to be sure that this is possible :)